Plugins~/Src/MeshSyncClientBlender/msblenContext.cpp

```diff
@@ -1635,6 +1634,130 @@ void msblenContext::WaitAndKickAsyncExport()
+    m_geometryNodeUtils.setInstancesDirty(true);
+#endif
+
+}
+
+/// Geometry Nodes Blender Context Functionality ///
+#if BLENDER_VERSION >= 300
+
+ms::InstanceInfoPtr msblenContext::exportInstanceInfo(
+    msblenContextState& state, 
+    msblenContextPathProvider& paths,  
+    Object* instancedObject, 
+    Object* parent, 
+    SharedVector<mu::float4x4> mat) {
+
+    auto info = ms::InstanceInfo::create();
+    info->path = paths.get_path(instancedObject);
+    info->parent_path = m_default_paths.get_path(parent); // parent will always be part of the scene
+
+    info->transforms = std::move(mat);
+
+    m_instances_manager.add(info);
+
+    return info;
+}
+
+void msblenContext::exportInstances() {
+   
+    bl::BlenderPyScene scene = bl::BlenderPyScene(bl::BlenderPyContext::get().scene());
+
+    std::unordered_set<Object*> scene_objects;
+    scene.each_objects([this, &scene_objects](Object* obj)
+        {
+            scene_objects.insert(obj);
+        });
+
+    // Assume everything is now dirty
+    m_instances_state->manager.setAlwaysMarkDirty(true);
+
+    m_geometryNodeUtils.each_instanced_object([this, &scene_objects](Object* instanced, Object* parent, SharedVector<mu::float4x4> matrices, bool fromFile){
+        // If the instanced object is not present in the file
+        if (!fromFile) {
+            return exportInstancesFromTree(instanced, parent, std::move(matrices));
+        }
+
+        // check if the object has been already exported as part of the scene
+        auto scene_object = scene_objects.find(instanced);
+        if (scene_object == scene_objects.end()) {
+            return exportInstacesFromFile(instanced, parent, std::move(matrices));
+        }
+        else {
+            return exportInstacesFromScene(instanced, parent, std::move(matrices));
+        }
+    });
+
+    m_geometryNodeUtils.setInstancesDirty(false);
+
+    scene_objects.clear();   
+}
+void msblenContext::exportInstacesFromFile(Object* instancedObject, Object* parent, SharedVector<mu::float4x4> mat)
+{
+    auto settings = m_settings;
+    settings.BakeTransform = false;
+
+    // There is some race condition that is causing rendering glitches on Unity. Seems related to UVs or triangle indices.
+    // Not using threads seems to fix it but should be investigated more.
+    settings.multithreaded = false;
+
+    auto transform = exportObject(*m_instances_state, m_default_paths, settings, instancedObject, false);
+
+    auto object_world_matrix = getWorldMatrix(instancedObject);
+    auto inverse = mu::invert(object_world_matrix);
+;
+    mu::parallel_for(0, mat.size(), 10, [this, &mat, &inverse](int i) 
+        {
+            mat[i] = m_geometryNodeUtils.blenderToUnityWorldMatrix(mat[i] * inverse);
+        });
+
+    exportInstanceInfo(*m_instances_state, m_default_paths, instancedObject, parent, std::move(mat));
+}
+void msblenContext::exportInstacesFromScene(Object* instancedObject, Object* parent, SharedVector<mu::float4x4> mat)
```

Comment:
Almost the same as `exportInstacesFromFile` ?
Maybe we can reduce code duplicity ?

Reply:
Removed duplicate code